Feature/update NestJs to version 11 (#4251)

the13-HK opened the pull request:
Sorry @dtslvr, 
I re-created the PR because I made a mistake in rebase and the commit history seemed to be messed up.
There were also multiple commits, so I squashed the commits.(Origin PR is https://github.com/ghostfolio/ghostfolio/pull/4265)


this PR is to resolve issue https://github.com/ghostfolio/ghostfolio/issues/4251. Please check
Confirmed that it passes by running npm test in local

Changes

Update NestJs to version 11
Update cache manager to version 6 due to Nest 11 requirements
Delete Import Milliseconds from cache-manager, because Milliseconds Type has been removed in version 6

the13-HK commented:
`apps/api/src/app/auth/api-key.strategy.ts` faced a problem as below.
Fixed it , and committed. 
https://github.com/nestjs/passport/issues/1857#issuecomment-2604035941

dtslvr commented on apps/api/src/app/auth/api-key.strategy.ts:
```diff
@@ -21,38 +21,38 @@ export class ApiKeyStrategy extends PassportStrategy(
-          done(null, user);
-        } catch (error) {
-          done(error, null);
+  public validate = async (
```
Could you use the function declaration without arrow syntax for consistency?

the13-HK replied:
@dtslvr 
Fixed. I was worried about whether to insert a line break before or after, but since there were many cases where a break was inserted before a conditional statement by referring to other files, I followed suit.
(I also think it's easier to see.)

dtslvr commented on apps/api/src/app/auth/api-key.strategy.ts:
```diff
@@ -21,38 +21,38 @@ export class ApiKeyStrategy extends PassportStrategy(
+  ) => {
+    try {
+      const user = await this.validateApiKey(apiKey);
+      if (this.configurationService.get('ENABLE_FEATURE_SUBSCRIPTION')) {
```
Please add a line break

the13-HK replied:
It's done!

dtslvr reviewed (changes requested):
Thanks for your contribution, @the13-HK! 😎
I have reviewed your changes and added a few comments.

dtslvr reviewed (changes requested):
Thanks for the adaptions.

When I start the server, I get the following error:
```
ERROR [ExceptionHandler] TypeError: Cannot read properties of undefined (reading 'client')
    at new RedisCacheService (webpack:///./src/app/redis-cache/redis-cache.service.ts:17:32)
    at Injector.instantiateClass (/development/ghostfolio/node_modules/@nestjs/core/injector/injector.js:373:19)
    at callback (/development/ghostfolio/node_modules/@nestjs/core/injector/injector.js:65:45)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Injector.resolveConstructorParams (/development/ghostfolio/node_modules/@nestjs/core/injector/injector.js:145:24)
    at async Injector.loadInstance (/development/ghostfolio/node_modules/@nestjs/core/injector/injector.js:70:13)
    at async Injector.loadProvider (/development/ghostfolio/node_modules/@nestjs/core/injector/injector.js:98:9)
    at async /development/ghostfolio/node_modules/@nestjs/core/injector/instance-loader.js:56:13
    at async Promise.all (index 3)
    at async InstanceLoader.createInstancesOfProviders (/development/ghostfolio/node_modules/@nestjs/core/injector/instance-loader.js:55:9)
```

dtslvr commented:
How are you progressing, @the13-HK?

the13-HK commented:
@dtslvr 
Sorry for late,,,,
As it is now I am stuck at a point where I cannot change the implementation of keys(prefix), which obtains a list of multiple keys using prefix as an argument.
I was looking for an improvement and realized a few days ago that it could be solved by using a namespace when declaring the keyv store, so I would like to try it this weekend.
I'm a little cautious because the structure of the method is likely to change.

dtslvr commented:
Thanks for the update, @the13-HK. Could you explain the reasoning behind replacing `cache-manager-redis-yet` with `@keyv/redis`?

the13-HK commented:
I haven't mentioned it here yet, @dtslvr 
With the migration of nestjs to version 11,  nestjs/cache-manager library has also been subject to a major update.
nestjs/cache-manager forces the use of keyv in cache-manager V6 (@nestjs/cache-manager 3.00 or higher).

There is also a statement below that asks you to change to keyv.

https://docs.nestjs.com/migration-guide#cache-module
https://github.com/nestjs/cache-manager/releases
https://www.npmjs.com/package/cache-manager-redis-yet

the13-HK commented:
Thank you for waiting, @dtslvr
The following two points were corrected to resolve the error.
- Regarding the processing of the getkeys method that returns a list of keys matching prefix, refer to the issue below and correct it to use iterator.
https://github.com/jaredwray/keyv/issues/40
- Since there was no suitable method for monitoring the life and death, I changed it to a method that monitors the life and death based on whether the option value (TTL) can be obtained. This is not a fundamental change, so there is room for improvement

I have confirmed that it can be operated without errors in the local environment and that tests do not result in errors.

dtslvr commented on apps/api/src/app/redis-cache/redis-cache.service.ts:
```diff
@@ -2,38 +2,34 @@ import { ConfigurationService } from '@ghostfolio/api/services/configuration/con
+    const prefix = aPrefix;
+    const keyList = [];
+    this.cache.stores[0].iterator((key) => {
+      if (prefix && key.startsWith(prefix)) {
```
What about this?
```
if ((prefix && key.startsWith(prefix)) || !prefix) {
```

the13-HK replied:
I have reflected the suggested corrections and removed the dead logic.

dtslvr commented on apps/api/src/app/redis-cache/redis-cache.service.ts:
```diff
@@ -2,38 +2,34 @@ import { ConfigurationService } from '@ghostfolio/api/services/configuration/con
-    }
+    const prefix = aPrefix;
+    const keyList = [];
+    this.cache.stores[0].iterator((key) => {
```
Please add a line break

the13-HK replied:
Added a line break

dtslvr commented on apps/api/src/app/redis-cache/redis-cache.service.ts:
```diff
@@ -62,10 +58,9 @@ export class RedisCacheService {
     try {
-      const client = this.cache.store.client;
-
+      const client = this.cache.stores[0];
```
Would it make sense to introduce a private getter?
```
private getStore() {
  return this.cache.stores[0];
}
```

Could also be used in `getKeys()`.

the13-HK replied:
I adopted getKeys() as the method for checking health checks.

dtslvr commented on package.json:
```diff
@@ -75,26 +75,27 @@
-    "@nestjs/schedule": "4.1.2",
-    "@nestjs/serve-static": "4.0.2",
-    "@prisma/client": "6.4.1",
+    "@keyv/redis": "^4.2.0",
```
Can you use a fixed version please: `"4.3.1"`

the13-HK replied:
I have updated the version

dtslvr reviewed (changes requested):
Thanks for your adaptions, @the13-HK! 😎
I have added a few more comments. Could you please rebase besides that? Thank you.

the13-HK commented:
Hi @dtslvr !
Sorry for the late response.
I responded to your comments.

dtslvr commented on apps/api/src/app/redis-cache/redis-cache.service.ts:
```diff
@@ -2,38 +2,33 @@ import { ConfigurationService } from '@ghostfolio/api/services/configuration/con
     return this.cache.get(key);
   }
 
   public async getKeys(aPrefix?: string): Promise<string[]> {
```
I'm seeing a warning here: `Async method 'getKeys' has no 'await' expression.`
This might be related to why I'm getting an empty keys array during testing.

Reproduction steps:
1. Navigate to _Accounts_
2. Click the _+_ button
3. Create a new account
4. Verify that `keys` is populated

I've attempted to fix this, but haven't been successful. @the13-HK, do you have any insights on why the keys array might be empty?

the13-HK replied:
Hi @dtslvr ,
> Verify that keys is populated
Can you tell me what you actually did in this step?
Did you set up a debug log or check the contents of Redis directly?

dtslvr replied:
I logged the result of `getKeys()` and it was empty.

the13-HK replied:
Hi @dtslvr 
I just fixed 2problems
- Keys can't be registered correctly in Redis
  - This was the same issue discussed in the following issue a few days ago.
    - https://github.com/jaredwray/keyv/issues/1349
  - The solution was presented in the NestJS discord community, and it was solved by changing the way redis was created in redis-cache.module.ts.
    - https://discord.com/channels/520622812742811698/1355619765804204163
- Iterator handling was not working properly.
  - This is the same as the issue mentioned in the following issue. The type of iterator() is AsyncGenerator, so the method of calling it has been changed. The argument to iterator() requires an argument to specify the namespace, but this is not necessary in this case, so it is not declared. 
  - This implementation method is discussed in the following issue and may change in future updates.
    - https://github.com/jaredwray/keyv/issues/1289

I confirmed that it was correctly registered in Redis and that the list of keys was correctly obtained using getkeys.

dtslvr replied:
Thanks again for your changes! I can confirm that `getKeys()` now returns a value!

However, when I import test activities (see [ok.json](https://github.com/ghostfolio/ghostfolio/blob/main/test/import/ok.json)), I get the following error at the same place and the application crashes!

```
/Users/dev/ghostfolio/node_modules/@keyv/serialize/dist/index.cjs:66
var defaultDeserialize = (data) => JSON.parse(data, (_, value) => {
                                        ^
SyntaxError: Unexpected number in JSON at position 1
    at JSON.parse (<anonymous>)
    at Keyv.defaultDeserialize [as _deserialize] (/Users/dev/ghostfolio/node_modules/@keyv/serialize/dist/index.cjs:66:41)
    at Keyv.deserializeData (/Users/dev/ghostfolio/node_modules/keyv/dist/index.cjs:764:19)
    at Keyv.function_ (/Users/dev/ghostfolio/node_modules/keyv/dist/index.cjs:442:33)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async RedisCacheService.getKeys (webpack:///./src/app/redis-cache/redis-cache.service.ts:25:26)
    at async RedisCacheService.removePortfolioSnapshotsByUserId (webpack:///./src/app/redis-cache/redis-cache.service.ts:85:18)
```

the13-HK replied:
@dtslvr 
This errors appears that data when iterating is determined to be JSON by default.
Ordinary strings will not result in an error, but hyphens used in snapshot IDs will cause an error.

The error can be avoided by disabling the default deserialize processing of the iterate process, but in the future, if we want to iterate over JSON data or value processing, it looks like we will need a custom deserialize function.
https://keyv.org/docs/keyv/#custom-serializers

jaredwray commented:
@the13-HK can you send me what the fix was for the disconnects? https://discord.com/channels/520622812742811698/1355619765804204163

I am working on a perm fix for this.

the13-HK commented on apps/api/src/app/redis-cache/redis-cache.service.ts:
```diff
@@ -2,38 +2,33 @@ import { ConfigurationService } from '@ghostfolio/api/services/configuration/con
-    const client = cache.store.client;
-
-    client.on('error', (error) => {
-      Logger.error(error, 'RedisCacheService');
```
This logic can keep it by using custom EventEmitter,
Fix it , but I haven't seen the output yet, so I'm still checking.
http://keyv.org/docs/keyv/#events

the13-HK commented:
@jaredwray 
Hi ! I've shared the fix on the discode server! I'll try to respond if there's any feedback.

dtslvr reviewed (changes requested):
Thanks again for all your efforts, @the13-HK! I have tested again and found another strange behavior. Here are the steps to reproduce:

1. Add `console.log({ portfolioSnapshotKey });` before `return portfolioSnapshotKey;` in `getPortfolioSnapshotKey()` of `RedisCacheService`
2. Click _Get Started_ to create a new user account
3. Navigate to _Portfolio_ > _Activities_ and import [ok.json](https://github.com/ghostfolio/ghostfolio/blob/main/test/import/ok.json)
4. Logout
5. Click _Get Started_ again to create a new user account
6. Navigate to _Portfolio_ > _Activities_ and import [ok.json](https://github.com/ghostfolio/ghostfolio/blob/main/test/import/ok.json)

You'll observe an infinite loop where the portfolio snapshot is recalculated repeatedly:
```
[Nest] 29505  - 04/12/2025, 9:48:03 AM   DEBUG [OrderService] Completed execution of getOrdersForPortfolioCalculator() in 0.079 seconds
[Nest] 29505  - 04/12/2025, 9:48:03 AM   DEBUG [AccountBalanceService] Completed execution of getAccountBalances() in 0.001 seconds
[Nest] 29505  - 04/12/2025, 9:48:03 AM   DEBUG [PortfolioCalculator] Completed execution of computeTransactionPoints() in 0.000 seconds
[Nest] 29505  - 04/12/2025, 9:48:03 AM   DEBUG [PortfolioCalculator] Completed execution of computeSnapshot() in 0.000 seconds
[Nest] 29505  - 04/12/2025, 9:48:03 AM     LOG [PortfolioSnapshotProcessor (PORTFOLIO)] Portfolio snapshot calculation of user 'b8558d49-175e-4aec-858c-62331ad06bdc' has been completed in 0.080 seconds
[Nest] 29505  - 04/12/2025, 9:48:03 AM     LOG [PortfolioSnapshotProcessor (PORTFOLIO)] Portfolio snapshot calculation of user 'b8558d49-175e-4aec-858c-62331ad06bdc' has been started
```

the13-HK commented:
@dtslvr 
I'm sorry for the repeated errors... I'll look into it.
 If there are any other steps you're planning to investigate, could you let me know what they are? 
I'll check those steps when I fix them.

dtslvr commented:
> @dtslvr I'm sorry for the repeated errors... I'll look into it. If there are any other steps you're planning to investigate, could you let me know what they are? I'll check those steps when I fix them.

No worries, I didn't expect this upgrade to be so tricky. However, there weren't many options for reducing the scope.
Maybe you could rebase this branch again.

the13-HK commented:
Hi @dtslvr 
I'm sorry for the late response.  I fixed it.
- The immediate cause was due to the processing specifications of keyv's deserialize.  
    - The key's defaultDeserialize uses a simple JSON.parse, but if the string contains a hyphen, it will not be converted properly and an error will occur.
https://keyv.org/docs/keyv/#optionsdeserialize
    - At first, I disabled the Deserialize process itself, but this caused problems, so I changed it to implement a Deserialize function that handles errors.

dtslvr reviewed (approved):
Thanks again @the13-HK and congrats on your first contribution! 🙌🏻
Ready to merge now.
